limiter.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"sort"
+	"strconv"
+	"time"
+)
+
+type (
+	// Config provides configuration values for creating a new rate-limiter.
+	Config struct {
+		// Redis is the Redis client for this rate-limiter.
+		Redis Eval
+
+		// Prefix is a string that will be added to the beginning of all keys.
+		Prefix string
+
+		// Backoff is the backoff scaling function (default 2x linear).
+		Backoff Backoff
+
+		// Buckets are the rate-limiting metrics.
+		Buckets []Bucket
+	}
+
+	// Limiter provides a single rate-limiter instance.
+	Limiter struct {
+		args    []any
+		redis   Eval
+		prefix  string
+		backoff Backoff
+	}
+
+	// Result provides the result of a rate-limiting test.
+	Result struct {
+		// Allow indicates whether the request should be allowed.
+		Allow bool
+
+		// Free indicates the remaining capacity before calls will be rejected.
+		Free float64
+
+		// Wait indicates how long the caller should wait before trying again.
+		Wait time.Duration
+	}
+)
+
+// NewLimiter creates a new rate-limiter instance.
+func NewLimiter(c Config) (*Limiter, error) {
+	if c.Redis == nil {
+		return nil, errors.New("the limiter must have a redis client")
+	}
+
+	if len(c.Buckets) == 0 {
+		return nil, errors.New("the limiter must have provided rates")
+	}
+
+	if c.Backoff == nil {
+		c.Backoff = Linear(2)
```

Comment:
if you use the functional options, you can actually hide away the kind of backoff from the public interface and just allow people to say `limiter.WithLinearBackoff(2)`, `limiter.WithExponentialBackoff(3)`, etc.